### ISSUE TITLE
Get filtered set of nodes in a subtree

### ISSUE THREAD

**Issue by sisp** [xaviergonz/mobx-keystone#247]
I'd like to be able to get a _computed_ set of all nodes in a subtree optionally filtered by some predicate. It seems [`computedWalkTreeAggregate`](https://github.com/xaviergonz/mobx-keystone/blob/9be6e0f386547c7b5a58eedd3566bbfa6a63fece/packages/lib/src/parent/walkTree.ts#L138-L161) is already doing the heavy lifting. Would you be open to adding such a function? E.g.:

```ts
function getNodes(subtreeRoot: object, predicate?: (node: object) => boolean): Set<object> {
  // ...
}
```

(I'm open to finding a better name :smile:.)

**Comment by sisp** [comment]
Regarding the name, maybe `findNodes` like [`findParent`](https://github.com/xaviergonz/mobx-keystone/blob/9be6e0f386547c7b5a58eedd3566bbfa6a63fece/packages/lib/src/parent/findParent.ts#L17)?

**Comment by sisp** [comment]
It seems that [`getChildrenObjects`](https://github.com/xaviergonz/mobx-keystone/blob/9be6e0f386547c7b5a58eedd3566bbfa6a63fece/packages/lib/src/parent/getChildrenObjects.ts#L12-L25) with `{ deep: true }` already returns a computed set of non-primitive nodes. Would it be better to use

1. `getChildrenObjects` and filter the returned set using the predicate, or to use
2. `computedWalkTreeAggregate` and convert the returned map to a set?

I _think_ `computedWalkTreeAggregate` might be more efficient because the predicate is applied in each nested computed value, which caches the result, whereas with `getChildrenObjects` any change in the subtree requires filtering the entire set of nodes returned by `getChildrenObjects`. What do you think?

**Comment by sisp** [comment]
I've submitted a PR with an attempt at closing this issue. I went for using `getChildrenObjects` because
1. it's simpler,
1. I think filtering a set should be quite cheap, and
1. `computedWalkTreeAggregate` requires a key for each entry in the returned map which is a model ID for root refs and `resolveId` but the key for e.g. an array node is not so obvious and I think `findChildren` (as a complement to `findParent`) should not be restricted to finding only model nodes.

Let me know what you think :slightly_smiling_face:
